opm/simulators/wells/BlackoilWellModel_impl.hpp

```diff
@@ -415,14 +421,17 @@ namespace Opm {
         updateWellTestState(simulationTime, wellTestState_);
 
         // calculate the well potentials for output
-        // TODO: when necessary
         try {
             std::vector<double> well_potentials;
             computeWellPotentials(well_potentials, local_deferredLogger);
         } catch ( std::runtime_error& e ) {
             const std::string msg = "A zero well potential is returned for output purposes. ";
             local_deferredLogger.warning("WELL_POTENTIAL_CALCULATION_FAILED", msg);
+        } catch (std::logic_error& e ) {
```

Comment:
Where can this exception be triggered?

Reply:
If the inverting of D fails it throws a numerical_error that is catched by the Deferred_logger that re-throw a logical_error in logAndCheckForExceptionsAndThrow. 

I don't know the details of the deferred_logger so I just added a fix for my problems. 
Maybe @atgeirr can explain why the exception is handled this way?

Reply:
Thanks for noticing!

The logAndCheckForExceptionsAndThrow() function does two things:
  - Writes all deferred log messages to the output rank.
  - If any rank have thrown an exception (and caught it, set the flag and passed to this function) then we throw logic_error.

I realize that this represents a loss of information compared to before for the serial case (however for the parallel case all we lost was race conditions...). Since the log messages are written, they should hopefully contain enough information to understand what happened, even if the higher level code no longer sees the exception type.

In order to re-throw the same type of exception, we would need to communicate the exception type somehow, and we have not made a facility for this. We could of course do that, it would requires us to define error codes (like in the good old days) to be used for this purpose and to change the catch spots. Perhaps adding an error code to the exception classes used also. It would be a bit of work to ensure it was done consistently across the whole system.

Reply:
It looks the `runtime_error` and `logic_error` are both related to the calculation of well potentials, I am wondering why this one is a `logic_error`? 

And since you are here, since when we terminate the simulation when we encounter `NumericalIssue`? Maybe it is just better to chop the time step to re-run to see whether the `NumericalIssue` is still there?